crates/ruff_python_formatter/src/statement/suite.rs

```diff
@@ -117,19 +108,55 @@ impl FormatRule<Suite, PyFormatContext<'_>> for FormatSuite {
                     }
                 });
 
-                joiner.entry_with_separator(&separator, &statement.format(), statement);
+                write!(f, [separator, statement.format()])?;
             } else {
-                joiner.entry(statement, &statement.format());
+                // Insert the appropriate number of empty lines based on the node level, e.g.:
+                // * [`NodeLevel::Module`]: Up to two empty lines
+                // * [`NodeLevel::CompoundStatement`]: Up to one empty line
+                // * [`NodeLevel::Expression`]: No empty lines
+                let separator = format_with(|f| {
+                    let count_lines = |offset| {
+                        // It's necessary to skip any trailing line comment because RustPython doesn't include trailing comments
+                        // in the node's range
+                        // ```python
+                        // a # The range of `a` ends right before this comment
+                        //
+                        // b
+                        // ```
+                        //
+                        // Simply using `lines_after` doesn't work if a statement has a trailing comment because
+                        // it then counts the lines between the statement and the trailing comment, which is
+                        // always 0. This is why it skips any trailing trivia (trivia that's on the same line)
+                        // and counts the lines after.
+                        let after_trailing_trivia = skip_trailing_trivia(offset, source);
+                        lines_after(after_trailing_trivia, source)
+                    };
+
+                    match node_level {
+                        NodeLevel::TopLevel => match count_lines(last.end()) {
+                            0 | 1 => hard_line_break().fmt(f),
+                            2 => empty_line().fmt(f),
+                            _ => write!(f, [empty_line(), empty_line()]),
+                        },
+                        NodeLevel::CompoundStatement => match count_lines(last.end()) {
+                            0 | 1 => hard_line_break().fmt(f),
+                            _ => empty_line().fmt(f),
+                        },
+                        NodeLevel::Expression(_) | NodeLevel::ParenthesizedExpression => {
+                            hard_line_break().fmt(f)
+                        }
+                    }
+                });
+
+                write!(f, [separator, statement.format()])?;
             }
 
             last = statement;
         }
 
-        let result = joiner.finish();
```

Comment:
Okay, I really need to write the restore context on drop helper... this new version never restored the context if any of the write operations above fail. This is not a real concern now but we should male sure to always properly restore the context. 

The easiest is to wrap the logic above in a format_with and call fmt directly on it

What I have in mind but is probably worth its own PR is to create a new struct that wraps a &mut Formatter<T>, implements Write, and restores the Context in its Drop implementation

Reply:
Ok got it, that makes sense. I will change it here.

Reply:
Actually I'm not sure I know how to do this. I wrapped the above in a `format_with`, but now I have:

```rust
formatted.fmt(f)?;

f.context_mut().set_node_level(saved_level);

Ok(())
```

Isn't that still problematic? This seems problematic too since the `format_with` would now use the wrong node level:

```rust
f.context_mut().set_node_level(saved_level);

formatted.fmt(f)
```

Reply:
You have to do:

```
let result = formatted.fmt(f); 
f.context_mut().set_node_level(saved_level); 
resut
```

Reply:
Ah right, ok, thank you.

Reply:
I can't do it because `iter` is captured in the closure :sob:

Reply:
Can move the `iter` into the closure too?

Reply:
Yes solved, sorry.